File: kotlin/bc-components/src/main/kotlin/com/blockchaincommons/bccomponents/EncryptedKey.kt
package com.blockchaincommons.bccomponents

import com.blockchaincommons.bctags.TAG_ENCRYPTED_KEY
import com.blockchaincommons.bcur.UR
import com.blockchaincommons.bcur.URCodable
import com.blockchaincommons.dcbor.Cbor
import com.blockchaincommons.dcbor.CborTaggedCodable
import com.blockchaincommons.dcbor.CborTaggedUtils
import com.blockchaincommons.dcbor.Tag
import com.blockchaincommons.dcbor.tagsForValues

/**
 * A symmetric content key encrypted using secret-based key derivation.
 *
 * [EncryptedKey] wraps an [EncryptedMessage] whose ciphertext is the
 * encrypted content key and whose AAD contains the CBOR-encoded key
 * derivation parameters. Multiple derivation methods are supported
 * (HKDF, PBKDF2, Scrypt, Argon2id).
 *
 * CDDL:
 * ```
 * EncryptedKey = #6.40027(EncryptedMessage)
 * ```
 *
 * Usage:
 * - Call [lock] to encrypt a content key with a chosen derivation method.
 * - Call [unlock] to recover the content key from a previously locked
 *   encrypted key.
 */
class EncryptedKey private constructor(
    private val params: KeyDerivationParams,
    private val encryptedMessage: EncryptedMessage,
) : CborTaggedCodable,
    URCodable {

    /** Returns the underlying encrypted message. */
    fun encryptedMessage(): EncryptedMessage = encryptedMessage

    /** Returns `true` if the derivation method is password-based. */
    fun isPasswordBased(): Boolean = params.isPasswordBased()

    /**
     * Decrypts this encrypted key using the given [secret] to recover the
     * original content key.
     *
     * The derivation parameters are read from the AAD of the encrypted
     * message, and the same key derivation is applied to produce the
     * decryption key.
     *
     * @throws BcComponentsException if decryption fails (e.g. wrong secret).
     */
    fun unlock(secret: ByteArray): SymmetricKey {
        val aad = encryptedMessage.aad()
        if (aad.isEmpty()) {
            throw BcComponentsException.general("Missing AAD CBOR in EncryptedMessage")
        }
        val paramsCbor = Cbor.tryFromData(aad)
        val a = paramsCbor.tryArray()
        if (a.isEmpty()) {
            throw BcComponentsException.general("Empty KeyDerivation array in AAD")
        }
        val index = a[0].tryInt()
        val method = KeyDerivationMethod.fromIndex(index)
            ?: throw BcComponentsException.general("Invalid KeyDerivationMethod index: $index")
        return when (method) {
            KeyDerivationMethod.HKDF ->
                HKDFParams.fromCbor(paramsCbor).unlock(encryptedMessage, secret)
            KeyDerivationMethod.PBKDF2 ->
                PBKDF2Params.fromCbor(paramsCbor).unlock(encryptedMessage, secret)
            KeyDerivationMethod.Scrypt ->
                ScryptParams.fromCbor(paramsCbor).unlock(encryptedMessage, secret)
            KeyDerivationMethod.Argon2id ->
                Argon2idParams.fromCbor(paramsCbor).unlock(encryptedMessage, secret)
        }
    }

    // -- toString --

    override fun toString(): String = "EncryptedKey($params)"

    // -- equals / hashCode --

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (other !is EncryptedKey) return false
        return params == other.params && encryptedMessage == other.encryptedMessage
    }

    override fun hashCode(): Int = 31 * params.hashCode() + encryptedMessage.hashCode()

    // -- CBOR --

    override fun cborTags(): List<Tag> = tagsForValues(listOf(TAG_ENCRYPTED_KEY))

    /**
     * The untagged CBOR representation is the tagged CBOR of the inner
     * [EncryptedMessage]. This means the wire format is:
     * `#6.40027(#6.40002([ciphertext, nonce, auth, aad]))`.
     */
    override fun untaggedCbor(): Cbor = encryptedMessage.taggedCbor()

    // -- Companion --

    companion object {
        /**
         * Encrypts a [contentKey] using the given key derivation [method]
         * and [secret].
         *
         * Default parameters are used for the chosen method.
         */
        fun lock(
            method: KeyDerivationMethod,
            secret: ByteArray,
            contentKey: SymmetricKey,
        ): EncryptedKey {
            val params = when (method) {
                KeyDerivationMethod.HKDF ->
                    KeyDerivationParams.HKDF(HKDFParams())
                KeyDerivationMethod.PBKDF2 ->
                    KeyDerivationParams.PBKDF2(PBKDF2Params())
                KeyDerivationMethod.Scrypt ->
                    KeyDerivationParams.Scrypt(ScryptParams())
                KeyDerivationMethod.Argon2id ->
                    KeyDerivationParams.Argon2id(Argon2idParams())
            }
            return lock(params, secret, contentKey)
        }

        /**
         * Encrypts a [contentKey] using the given key derivation [params]
         * and [secret].
         *
         * This overload allows specifying custom derivation parameters.
         */
        fun lock(
            params: KeyDerivationParams,
            secret: ByteArray,
            contentKey: SymmetricKey,
        ): EncryptedKey {
            val encryptedMessage = params.lock(contentKey, secret)
            return EncryptedKey(params, encryptedMessage)
        }

        /** Decodes an [EncryptedKey] from untagged CBOR. */
        fun fromUntaggedCbor(cbor: Cbor): EncryptedKey {
            val encryptedMessage = EncryptedMessage.fromTaggedCbor(cbor)
            val aad = encryptedMessage.aad()
            if (aad.isEmpty()) {
                throw BcComponentsException.general("Missing AAD in EncryptedKey")
            }
            val paramsCbor = Cbor.tryFromData(aad)
            val params = KeyDerivationParams.fromCbor(paramsCbor)
            return EncryptedKey(params, encryptedMessage)
        }

        /** Decodes an [EncryptedKey] from tagged CBOR. */
        fun fromTaggedCbor(cbor: Cbor): EncryptedKey =
            CborTaggedUtils.fromTaggedCbor(
                cbor,
                tagsForValues(listOf(TAG_ENCRYPTED_KEY)),
            ) { fromUntaggedCbor(it) }

        /** Decodes an [EncryptedKey] from tagged CBOR binary data. */
        fun fromTaggedCborData(data: ByteArray): EncryptedKey =
            CborTaggedUtils.fromTaggedCborData(
                data,
                tagsForValues(listOf(TAG_ENCRYPTED_KEY)),
            ) { fromUntaggedCbor(it) }

        /** Decodes an [EncryptedKey] from a UR. */
        fun fromUr(ur: UR): EncryptedKey {
            ur.checkType("encrypted-key")
            return fromUntaggedCbor(ur.cbor)
        }

        /** Decodes an [EncryptedKey] from a UR string. */
        fun fromUrString(urString: String): EncryptedKey =
            fromUr(UR.fromUrString(urString))
    }
}

File: csharp/BCComponents/BCComponents/EncryptedKey.cs
using BlockchainCommons.BCTags;
using BlockchainCommons.DCbor;

namespace BlockchainCommons.BCComponents;

/// <summary>
/// A symmetric content key encrypted using secret-based key derivation.
/// </summary>
/// <remarks>
/// <para>
/// <see cref="EncryptedKey"/> wraps an <see cref="EncryptedMessage"/> whose
/// ciphertext is the encrypted content key and whose AAD contains the
/// CBOR-encoded key derivation parameters. Multiple derivation methods are
/// supported (HKDF, PBKDF2, Scrypt, Argon2id).
/// </para>
/// <para>
/// CDDL:
/// <code>
/// EncryptedKey = #6.40027(EncryptedMessage)
/// </code>
/// </para>
/// </remarks>
public sealed class EncryptedKey : IEquatable<EncryptedKey>, ICborTaggedEncodable, ICborTaggedDecodable
{
    private readonly KeyDerivationParams _params;
    private readonly EncryptedMessage _encryptedMessage;

    private EncryptedKey(KeyDerivationParams @params, EncryptedMessage encryptedMessage)
    {
        _params = @params;
        _encryptedMessage = encryptedMessage;
    }

    /// <summary>Returns the underlying encrypted message.</summary>
    public EncryptedMessage EncryptedMessage => _encryptedMessage;

    /// <summary>Returns <c>true</c> if the derivation method is password-based.</summary>
    public bool IsPasswordBased => _params.IsPasswordBased;

    /// <summary>Returns <c>true</c> if the derivation method is SSH agent-based.</summary>
    public bool IsSshAgent => _params.IsSshAgent;

    /// <summary>
    /// Encrypts a content key using the given key derivation method and secret
    /// (default parameters).
    /// </summary>
    /// <param name="method">The key derivation method to use.</param>
    /// <param name="secret">The secret to derive the encryption key from.</param>
    /// <param name="contentKey">The content key to encrypt.</param>
    /// <returns>A new <see cref="EncryptedKey"/>.</returns>
    public static EncryptedKey Lock(
        KeyDerivationMethod method,
        byte[] secret,
        SymmetricKey contentKey)
    {
        var @params = method switch
        {
            KeyDerivationMethod.HKDF => KeyDerivationParams.FromHKDF(new HKDFParams()),
            KeyDerivationMethod.PBKDF2 => KeyDerivationParams.FromPBKDF2(new PBKDF2Params()),
            KeyDerivationMethod.Scrypt => KeyDerivationParams.FromScrypt(new ScryptParams()),
            KeyDerivationMethod.Argon2id => KeyDerivationParams.FromArgon2id(new Argon2idParams()),
            _ => throw BCComponentsException.General($"Unknown key derivation method: {method}"),
        };
        return Lock(@params, secret, contentKey);
    }

    /// <summary>
    /// Encrypts a content key using the given key derivation parameters and secret.
    /// </summary>
    /// <param name="params">The key derivation parameters.</param>
    /// <param name="secret">The secret to derive the encryption key from.</param>
    /// <param name="contentKey">The content key to encrypt.</param>
    /// <returns>A new <see cref="EncryptedKey"/>.</returns>
    public static EncryptedKey Lock(
        KeyDerivationParams @params,
        byte[] secret,
        SymmetricKey contentKey)
    {
        var encryptedMessage = @params.Lock(contentKey, secret);
        return new EncryptedKey(@params, encryptedMessage);
    }

    /// <summary>
    /// Decrypts this encrypted key using the given secret to recover the
    /// original content key.
    /// </summary>
    /// <param name="secret">The secret to derive the decryption key from.</param>
    /// <returns>The recovered <see cref="SymmetricKey"/>.</returns>
    /// <exception cref="BCComponentsException">
    /// Thrown if decryption fails (e.g. wrong secret).
    /// </exception>
    public SymmetricKey Unlock(byte[] secret)
    {
        var aad = _encryptedMessage.Aad;
        if (aad.Length == 0)
            throw BCComponentsException.General("Missing AAD CBOR in EncryptedMessage");

        var paramsCbor = Cbor.TryFromData(aad);
        var a = paramsCbor.TryIntoArray();
        if (a.Count == 0)
            throw BCComponentsException.General("Empty KeyDerivation array in AAD");

        var index = (int)a[0].TryIntoUInt64();
        var method = KeyDerivationMethodExtensions.FromIndex(index)
            ?? throw BCComponentsException.General($"Invalid KeyDerivationMethod index: {index}");

        return method switch
        {
            KeyDerivationMethod.HKDF => HKDFParams.FromCbor(paramsCbor).Unlock(_encryptedMessage, secret),
            KeyDerivationMethod.PBKDF2 => PBKDF2Params.FromCbor(paramsCbor).Unlock(_encryptedMessage, secret),
            KeyDerivationMethod.Scrypt => ScryptParams.FromCbor(paramsCbor).Unlock(_encryptedMessage, secret),
            KeyDerivationMethod.Argon2id => Argon2idParams.FromCbor(paramsCbor).Unlock(_encryptedMessage, secret),
            _ => throw new InvalidOperationException(),
        };
    }

    // --- IEquatable<EncryptedKey> ---

    /// <inheritdoc/>
    public bool Equals(EncryptedKey? other)
    {
        if (other is null) return false;
        return _params.Equals(other._params)
            && _encryptedMessage == other._encryptedMessage;
    }

    /// <inheritdoc/>
    public override bool Equals(object? obj) => obj is EncryptedKey k && Equals(k);

    /// <inheritdoc/>
    public override int GetHashCode() => HashCode.Combine(_params, _encryptedMessage);

    /// <summary>Tests equality of two EncryptedKey instances.</summary>
    public static bool operator ==(EncryptedKey? left, EncryptedKey? right) =>
        left is null ? right is null : left.Equals(right);

    /// <summary>Tests inequality of two EncryptedKey instances.</summary>
    public static bool operator !=(EncryptedKey? left, EncryptedKey? right) => !(left == right);

    // --- ICborTagged ---

    /// <summary>Returns the CBOR tags for this type (40027).</summary>
    public static IReadOnlyList<Tag> CborTags =>
        GlobalTags.TagsForValues(BcTags.TagEncryptedKey);

    // --- ICborTaggedEncodable ---

    /// <summary>
    /// Returns the untagged CBOR representation.
    /// </summary>
    /// <remarks>
    /// The untagged CBOR is the tagged CBOR of the inner <see cref="EncryptedMessage"/>.
    /// This means the wire format is:
    /// <c>#6.40027(#6.40002([ciphertext, nonce, auth, aad]))</c>.
    /// </remarks>
    public Cbor UntaggedCbor() => _encryptedMessage.TaggedCbor();

    /// <summary>Returns the tagged CBOR representation.</summary>
    public Cbor TaggedCbor() => Cbor.ToTaggedValue(CborTags[0], UntaggedCbor());

    /// <summary>Returns the tagged CBOR representation.</summary>
    public Cbor ToCbor() => TaggedCbor();

    // --- ICborTaggedDecodable ---

    /// <summary>
    /// Decodes an <see cref="EncryptedKey"/> from untagged CBOR.
    /// </summary>
    /// <param name="cbor">The untagged CBOR value (a tagged EncryptedMessage).</param>
    /// <returns>A new <see cref="EncryptedKey"/>.</returns>
    public static EncryptedKey FromUntaggedCbor(Cbor cbor)
    {
        var encryptedMessage = EncryptedMessage.FromTaggedCbor(cbor);
        var aad = encryptedMessage.Aad;
        if (aad.Length == 0)
            throw BCComponentsException.General("Missing AAD in EncryptedKey");

        var paramsCbor = Cbor.TryFromData(aad);
        var @params = KeyDerivationParams.FromCbor(paramsCbor);
        return new EncryptedKey(@params, encryptedMessage);
    }

    /// <summary>
    /// Decodes an <see cref="EncryptedKey"/> from tagged CBOR.
    /// </summary>
    /// <param name="cbor">The tagged CBOR value.</param>
    /// <returns>A new <see cref="EncryptedKey"/>.</returns>
    public static EncryptedKey FromTaggedCbor(Cbor cbor)
    {
        foreach (var tag in CborTags)
        {
            try
            {
                var item = cbor.TryIntoExpectedTaggedValue(tag);
                return FromUntaggedCbor(item);
            }
            catch (CborWrongTagException) { }
            catch (CborWrongTypeException) { }
        }
        throw new CborWrongTypeException();
    }

    // --- Display ---

    /// <inheritdoc/>
    public override string ToString() => $"EncryptedKey({_params})";
}
